File: web/src/test/java/no/nav/sbl/dialogarena/sikkerhet/CORSFilterTest.java
package no.nav.sbl.dialogarena.sikkerhet;

import org.junit.After;
import org.junit.Assert;
import org.junit.Test;
import org.springframework.mock.web.MockHttpServletResponse;


public class CORSFilterTest {

    private final CORSFilter corsFilter = new CORSFilter();
    private final String unknownOrigin = "https://www.unknown.no";

    @After
    public void tearDown() {
        System.clearProperty("environment.name");
    }

    @Test
    public void setCorsHeaders_inProdWithUnknownOrigin_shouldNotSetCorsHeaders() {
        MockHttpServletResponse httpResponse = new MockHttpServletResponse();
        corsFilter.setCorsHeaders(httpResponse, unknownOrigin);
        Assert.assertEquals(0, httpResponse.getHeaderNames().size());
    }

    @Test
    public void setCorsHeaders_inProdWithTrustedOrigin_shouldSetCorsHeaders() {
        MockHttpServletResponse httpResponse = new MockHttpServletResponse();

        String trustedOrigin = "https://www.nav.no";
        corsFilter.setCorsHeaders(httpResponse, trustedOrigin);

        Assert.assertTrue(httpResponse.containsHeader("Access-Control-Allow-Headers"));
        Assert.assertTrue(httpResponse.containsHeader("Access-Control-Allow-Methods"));
        Assert.assertTrue(httpResponse.containsHeader("Access-Control-Allow-Credentials"));
        Assert.assertEquals(trustedOrigin, httpResponse.getHeader("Access-Control-Allow-Origin"));
    }

    @Test
    public void setCorsHeaders_inTestWithUnknownOrigin_shouldSetCorsHeaders() {
        System.setProperty("environment.name", "q0");
        MockHttpServletResponse httpResponse = new MockHttpServletResponse();

        corsFilter.setCorsHeaders(httpResponse, unknownOrigin);

        Assert.assertTrue(httpResponse.containsHeader("Access-Control-Allow-Headers"));
        Assert.assertTrue(httpResponse.containsHeader("Access-Control-Allow-Methods"));
        Assert.assertTrue(httpResponse.containsHeader("Access-Control-Allow-Credentials"));
        Assert.assertEquals(unknownOrigin, httpResponse.getHeader("Access-Control-Allow-Origin"));
    }
}

File: src/test/kotlin/no/nav/sosialhjelp/soknad/app/filter/CORSFilterTest.kt
package no.nav.sosialhjelp.soknad.app.filter

import io.mockk.every
import io.mockk.mockk
import io.mockk.mockkObject
import io.mockk.unmockkObject
import jakarta.servlet.http.HttpServletRequest
import no.nav.sosialhjelp.soknad.app.MiljoUtils
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.AfterEach
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.springframework.mock.web.MockFilterChain
import org.springframework.mock.web.MockHttpServletResponse

internal class CORSFilterTest {

    private val corsFilter = CORSFilter()

    private val filterChain = MockFilterChain()

    @BeforeEach
    fun setUp() {
        mockkObject(MiljoUtils)
    }

    @AfterEach
    internal fun tearDown() {
        unmockkObject(MiljoUtils)
    }

    @Test
    fun setCorsHeaders_inProdWithUnknownOrigin_shouldNotSetCorsHeaders() {
        every { MiljoUtils.isNonProduction() } returns false

        val unknownOrigin = "https://www.unknown.no"

        val request: HttpServletRequest = mockk()
        every { request.requestURI } returns "requestUri"
        every { request.getHeader("Origin") } returns unknownOrigin

        val response = MockHttpServletResponse()

        corsFilter.doFilter(request, response, filterChain)

        assertThat(response.headerNames).isEmpty()
    }

    @Test
    fun setCorsHeaders_inProdWithTrustedOrigin_shouldSetCorsHeaders() {
        every { MiljoUtils.isNonProduction() } returns true
        val trustedOrigin = "https://www.nav.no"

        val request: HttpServletRequest = mockk()
        every { request.requestURI } returns "requestUri"
        every { request.getHeader("Origin") } returns trustedOrigin

        val response = MockHttpServletResponse()

        corsFilter.doFilter(request, response, filterChain)
        assertThat(response.headerNames).contains("Access-Control-Allow-Headers")
        assertThat(response.headerNames).contains("Access-Control-Allow-Methods")
        assertThat(response.headerNames).contains("Access-Control-Allow-Credentials")
        assertThat(response.getHeader("Access-Control-Allow-Origin")).isEqualTo(trustedOrigin)
    }

    @Test
    fun setCorsHeaders_inTestWithUnknownOrigin_shouldSetCorsHeaders() {
        every { MiljoUtils.isNonProduction() } returns true
        val unknownOrigin = "https://www.unknown.no"

        val request: HttpServletRequest = mockk()
        every { request.requestURI } returns "requestUri"
        every { request.getHeader("Origin") } returns unknownOrigin

        val response = MockHttpServletResponse()

        corsFilter.doFilter(request, response, filterChain)

        assertThat(response.headerNames).contains("Access-Control-Allow-Headers")
        assertThat(response.headerNames).contains("Access-Control-Allow-Methods")
        assertThat(response.headerNames).contains("Access-Control-Allow-Credentials")
        assertThat(response.getHeader("Access-Control-Allow-Origin")).isEqualTo(unknownOrigin)
    }
}
